File: tools/port-game-state.js
// port-game-state.js
// Transforms legacy game-state.js into TypeScript (game-state.ts)
// Run: node tools/port-game-state.js

const fs = require('fs');
const path = require('path');

const srcPath = path.join(__dirname, '../legacy-pure-js/js/game/game-state.js');
const dstPath = path.join(__dirname, '../src/engine/game-state.ts');

const src = fs.readFileSync(srcPath, 'utf8');

const output = [];

// Header
output.push('// @ts-nocheck');
output.push('// game-state.ts — Ported from legacy game-state.js');
output.push('// DO NOT add type annotations — ts-nocheck handles this file');
output.push('');
output.push("import * as Cards from './cards';");
output.push("import * as Mana from './mana';");
output.push("import * as Stack from './stack';");
output.push("import * as GameAI from './game-ai';");
output.push("import * as CombatSim from './combat-sim';");
output.push('');
output.push("export const PHASES = ['untap', 'upkeep', 'draw', 'main1', 'combat_begin', 'combat_attackers', 'combat_blockers', 'combat_damage', 'combat_end', 'main2', 'end', 'cleanup'];");
output.push('export const STARTING_LIFE = 20;');
output.push('export const MAX_HAND_SIZE = 7;');
output.push('');

function applyTransforms(line) {
  // Replace module references
  line = line.replace(/\bCardEngine\./g, 'Cards.');
  line = line.replace(/\bManaSystem\./g, 'Mana.');
  line = line.replace(/\bGameStack\./g, 'Stack.');
  line = line.replace(/\bCombatSystem\./g, 'CombatSim.');

  // Replace this.PHASES, this.STARTING_LIFE, this.MAX_HAND_SIZE
  line = line.replace(/\bthis\.PHASES\b/g, 'PHASES');
  line = line.replace(/\bthis\.STARTING_LIFE\b/g, 'STARTING_LIFE');
  line = line.replace(/\bthis\.MAX_HAND_SIZE\b/g, 'MAX_HAND_SIZE');

  // Replace all this.methodName( with methodName(
  const methodNames = [
    '_resolveSimpleEffect', '_checkTriggerCondition', '_checkEffectCondition',
    'fireTrigger', 'castSpell', 'creatureDies', 'sacrifice', 'advancePhase',
    '_processPhase', '_endOfTurnCleanup', '_checkWinner', '_createPlayer',
    '_registerTriggersForBattlefield', '_registerCardTriggers', '_unregisterCardTriggers',
    '_aiMulliganDecision', '_aiChooseTargetsForEffects', 'mulligan', 'keepHand',
    'startGame', '_updateDynamicStaticAbilities', '_updateDynamicPower',
    '_updateVividCreatures', '_applyStaticOnETB', 'playLand', '_processHideaway',
    '_checkHideawayCondition', 'activateHideaway', 'tapLandForMana',
    'calculateAffinityDiscount', 'getAttackingCreatureDiscount', 'returnTemporaryExiles',
    '_smartActivateManaAbilities', '_pickMostNeededColor', '_canUseAbility',
    '_getAbilityManaProduction', 'autoTapForSpell', '_findBestHybridCombo',
    'previewAutoTap', 'activateLoyaltyAbility', '_checkPlaneswalkerDeath',
    'damagePlaneswalker', '_applyAuraEffects', '_removeAuraEffects',
    'equipCreature', '_applyEquipmentEffects', '_removeEquipmentEffects',
    'transformCreature', '_advanceSagaChapter', 'activateCycling',
    'getPlayableCards', 'hasAffordableAbilities', 'getHarmonizableCards',
    'castHarmonize', 'submitExileChoice', '_hasValidTargetsForCard',
    '_effectRequiresTargets', '_hasValidTargetsForEffect', '_oracleTextRequiresTargets',
    '_hasValidTargetsFromOracle', '_processStateBasedActions', 'resolveTriggerCost',
    'getEffectiveCmcWithReduction', '_processSuspendedSpells', '_performBlight',
    '_setupBlightChoice', '_setupHandExileChoice', 'resolveHandExileChoice',
    'resolveManaChoice', 'resolveUnlessPay', 'resolveEndureChoice',
    'resolveMillLandChoice', 'resolveBlightChoice', 'resolveBuffChoice',
    'resolveMultiBuffChoice'
  ];

  for (const name of methodNames) {
    const re = new RegExp(`\\bthis\\.${name.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}\\(`, 'g');
    line = line.replace(re, `${name}(`);
  }

  // Remove VFX/UIGame calls
  // Pattern 1: if (typeof VFX !== 'undefined') VFX.xxx(...);
  line = line.replace(/if\s*\(typeof\s+VFX[^;]*;/g, '// TODO: VFX callback');
  // Pattern 2: VFX.xxx(...);
  line = line.replace(/\bVFX\.[a-zA-Z_]+\([^)]*\);?/g, '// TODO: VFX callback');
  // Pattern 3: UIGame.xxx(...);
  line = line.replace(/\bUIGame\.[a-zA-Z_]+\([^)]*\);?/g, '// TODO: UIGame callback');

  return line;
}

function countBraces(line) {
  let open = 0, close = 0;
  let inStr = false, strChar = '';
  let escaped = false;
  let inTempl = 0; // template literal depth
  for (let ci = 0; ci < line.length; ci++) {
    const ch = line[ci];
    if (escaped) { escaped = false; continue; }
    if (ch === '\\') { escaped = true; continue; }
    if (inStr && strChar !== '`') {
      if (ch === strChar) inStr = false;
      continue;
    }
    if (inStr && strChar === '`') {
      if (ch === '`') inStr = false;
      // Don't count braces inside template literals unless in ${...}
      // Simplified: just skip
      continue;
    }
    if (!inStr) {
      if (ch === '"' || ch === "'") { inStr = true; strChar = ch; continue; }
      if (ch === '`') { inStr = true; strChar = ch; continue; }
      if (ch === '/') {
        // Check for line comment
        if (ci + 1 < line.length && line[ci + 1] === '/') break; // rest is comment
        // Skip regex detection (complex) - just continue
      }
      if (ch === '{') open++;
      if (ch === '}') close++;
    }
  }
  return { open, close };
}

const methodNames = [
  'create', '_aiMulliganDecision', '_aiChooseTargetsForEffects', 'mulligan', 'keepHand',
  'startGame', '_registerTriggersForBattlefield', '_registerCardTriggers', '_unregisterCardTriggers',
  'fireTrigger', '_checkTriggerCondition', '_checkEffectCondition', '_resolveSimpleEffect',
  '_processSuspendedSpells', '_performBlight', '_setupBlightChoice', '_setupHandExileChoice',
  'resolveHandExileChoice', 'resolveManaChoice', 'resolveUnlessPay', 'resolveEndureChoice',
  'resolveMillLandChoice', 'resolveBlightChoice', 'resolveBuffChoice', 'resolveMultiBuffChoice',
  '_createPlayer', 'advancePhase', '_processPhase', 'playLand', '_processHideaway',
  '_checkHideawayCondition', 'activateHideaway', 'tapLandForMana', 'calculateAffinityDiscount',
  'getAttackingCreatureDiscount', 'returnTemporaryExiles', 'castSpell', '_smartActivateManaAbilities',
  '_pickMostNeededColor', '_canUseAbility', '_getAbilityManaProduction', 'autoTapForSpell',
  '_findBestHybridCombo', 'previewAutoTap', '_updateVividCreatures', '_updateDynamicPower',
  '_updateDynamicStaticAbilities', '_applyStaticOnETB', '_endOfTurnCleanup', '_checkWinner',
  'activateLoyaltyAbility', '_checkPlaneswalkerDeath', 'damagePlaneswalker', '_applyAuraEffects',
  '_removeAuraEffects', 'equipCreature', '_applyEquipmentEffects', '_removeEquipmentEffects',
  'creatureDies', 'transformCreature', 'sacrifice', '_advanceSagaChapter', 'activateCycling',
  'getPlayableCards', 'hasAffordableAbilities', 'getHarmonizableCards', 'castHarmonize',
  'submitExileChoice', '_hasValidTargetsForCard', '_effectRequiresTargets', '_hasValidTargetsForEffect',
  '_oracleTextRequiresTargets', '_hasValidTargetsFromOracle', '_processStateBasedActions',
  'resolveTriggerCost', 'getEffectiveCmcWithReduction'
];

function detectMethodStart(line, relDepth) {
  if (relDepth !== 0) return null;
  const trimmed = line.trim();
  for (const name of methodNames) {
    if (trimmed.startsWith(name + '(') || trimmed.startsWith('async ' + name + '(')) {
      return name;
    }
    // Also match with comment before: "// === Comment ===\n  methodName("
    // No, trim handles leading spaces
  }
  return null;
}

const allLines = src.split('\n');
let relDepth = 0;
let methods = [];
let currentM = null;

for (let li = 1; li < allLines.length; li++) {
  const rawLine = allLines[li];
  const bc = countBraces(rawLine);

  if (currentM === null) {
    const mName = detectMethodStart(rawLine, relDepth);
    if (mName) {
      currentM = { name: mName, lines: [] };
      currentM.lines.push(rawLine);
      relDepth += bc.open - bc.close;
    } else {
      relDepth += bc.open - bc.close;
      if (relDepth < 0) break;
    }
  } else {
    currentM.lines.push(rawLine);
    relDepth += bc.open - bc.close;
    if (relDepth === 0) {
      methods.push({ ...currentM });
      currentM = null;
    }
  }
}

console.log(`Found ${methods.length} methods`);

for (const method of methods) {
  const lines = method.lines;
  let firstLine = lines[0].trim();

  let funcName = method.name;
  if (funcName === 'create') funcName = 'createGame';

  if (firstLine.startsWith('async ')) {
    firstLine = firstLine.replace(/^async (\w+)\(/, `export async function ${funcName}(`);
  } else {
    firstLine = firstLine.replace(/^(\w+)\(/, `export function ${funcName}(`);
  }

  firstLine = applyTransforms(firstLine);
  output.push(firstLine);

  for (let li = 1; li < lines.length - 1; li++) {
    output.push(applyTransforms(lines[li]));
  }

  const lastLine = lines[lines.length - 1];
  const lastTrimmed = lastLine.trim();
  if (lastTrimmed === '},' || lastTrimmed === '}') {
    output.push('}');
  } else {
    output.push(applyTransforms(lastLine));
  }

  output.push('');
}

const result = output.join('\n');
fs.writeFileSync(dstPath, result);
console.log(`Written ${result.split('\\n').length} lines to ${dstPath}`);

File: tools/port-game-state.py
#!/usr/bin/env python3
"""Port legacy game-state.js to TypeScript game-state.ts"""
import re

src_path = 'C:/Users/lucas/OneDrive/Área de Trabalho/magic_draft/legacy-pure-js/js/game/game-state.js'
dst_path = 'C:/Users/lucas/OneDrive/Área de Trabalho/magic_draft/src/engine/game-state.ts'

with open(src_path, 'r', encoding='utf-8') as f:
    content = f.read()

lines = content.split('\n')
print(f'Source: {len(lines)} lines')

output = []

# Header
output.append('// @ts-nocheck')
output.append('// game-state.ts — Ported from legacy game-state.js')
output.append('// DO NOT add type annotations — ts-nocheck handles this file')
output.append('')
output.append("import * as Cards from './cards';")
output.append("import * as Mana from './mana';")
output.append("import * as CombatSim from './combat-sim';")
output.append("import * as CardUtils from './card-utils';")
output.append('')

in_gamestate_obj = False
brace_depth = 0
found_gamestate = False

for i, line in enumerate(lines):
    # Skip original header/variable declaration, replace with exports
    stripped = line.strip()

    # Detect GameState object start
    if not found_gamestate and ('const GameState' in line or 'var GameState' in line or 'let GameState' in line) and '=' in line and '{' in line:
        found_gamestate = True
        in_gamestate_obj = True
        brace_depth = line.count('{') - line.count('}')
        # Skip this line - we'll handle methods individually
        output.append('// ─── GameState methods (converted to exported functions) ───')
        continue

    if in_gamestate_obj:
        # Count braces to track when we exit the object
        brace_depth += line.count('{') - line.count('}')

        if brace_depth <= 0:
            # End of GameState object
            in_gamestate_obj = False
            output.append('// ─── End of GameState object ───')
            continue

        # Convert method definitions: "  methodName(args) {" -> "export function methodName(args) {"
        method_match = re.match(r'^(\s+)(\w+)\s*\(([^)]*)\)\s*\{', line)
        if method_match and brace_depth == 1 + line.count('{') - line.count('}'):
            indent = method_match.group(1)
            method_name = method_match.group(2)
            args = method_match.group(3)
            rest = line[line.index('{'):]

            # Replace this.X -> X (for GameState methods calling each other)
            new_line = f'export function {method_name}({args}) {rest}'
            output.append(new_line)
            continue

        # Replace this. references to GameState methods - they become direct calls
        line = re.sub(r'\bthis\.(?!_uid|_zone|_tapped|_powerMod|_toughnessMod|_counters|_damage|_summoningSick|_attacking|_blocking|_triggers|_tempKeywords|_champion|_evoked|_attachedTo|_attachments|_isSaga|_sagaChapter|_sagaMaxChapter|_loyaltyUsedThisTurn|_regenerateShield|_stun|power|toughness|cmc|mana_cost|name|type_line|oracle_text|colors|color_identity|rarity|id)', '', line)

        # Remove VFX and UIGame calls
        line = re.sub(r'\bVFX\.\w+\([^;]*\);?', '// VFX removed', line)
        line = re.sub(r'\bUIGame\.\w+\([^;]*\);?', '// UIGame removed', line)

        output.append(line)
    else:
        # Outside GameState object - keep as-is but do some replacements
        # Replace CardEngine. -> Cards.
        line = line.replace('CardEngine.', 'Cards.')
        # Replace ManaSystem. -> Mana.
        line = line.replace('ManaSystem.', 'Mana.')
        # Replace CombatSystem. -> CombatSim.
        line = line.replace('CombatSystem.', 'CombatSim.')

        # Remove VFX and UIGame
        line = re.sub(r'\bVFX\.\w+\([^;]*\);?', '// VFX removed', line)
        line = re.sub(r'\bUIGame\.\w+\([^;]*\);?', '// UIGame removed', line)

        output.append(line)

result = '\n'.join(output)
with open(dst_path, 'w', encoding='utf-8') as f:
    f.write(result)

out_lines = len(output)
print(f'Output: {out_lines} lines -> {dst_path}')
print('Done!')
